File: pkg/geometry/triangle.go
package geometry

import (
	"math"

	"sif/gogs/eric/relray/pkg/vec"
)

// Triangle is defined by three vertices. Uses the Moller-Trumbore algorithm.
type Triangle struct {
	V0, V1, V2 vec.Vec3
}

func (tri *Triangle) Intersect(origin, dir vec.Vec3, tMin, tMax float64) (Hit, bool) {
	edge1 := tri.V1.Sub(tri.V0)
	edge2 := tri.V2.Sub(tri.V0)
	h := dir.Cross(edge2)
	a := edge1.Dot(h)

	if math.Abs(a) < 1e-12 {
		return Hit{}, false // ray parallel to triangle
	}

	f := 1.0 / a
	s := origin.Sub(tri.V0)
	u := f * s.Dot(h)
	if u < 0 || u > 1 {
		return Hit{}, false
	}

	q := s.Cross(edge1)
	v := f * dir.Dot(q)
	if v < 0 || u+v > 1 {
		return Hit{}, false
	}

	t := f * edge2.Dot(q)
	if t < tMin || t > tMax {
		return Hit{}, false
	}

	p := origin.Add(dir.Scale(t))
	normal := edge1.Cross(edge2).Normalize()
	hit := Hit{T: t, Point: p}
	hit.SetFaceNormal(dir, normal)
	return hit, true
}

File: rust/src/geometry/triangle.rs
use crate::vec::Vec3;
use super::{Hit, Shape};

pub struct Triangle {
    pub v0: Vec3,
    pub v1: Vec3,
    pub v2: Vec3,
}

impl Triangle {
    pub fn intersect_tri(
        &self,
        origin: Vec3,
        dir: Vec3,
        t_min: f64,
        t_max: f64,
    ) -> Option<Hit> {
        let edge1 = self.v1 - self.v0;
        let edge2 = self.v2 - self.v0;
        let h = dir.cross(edge2);
        let a = edge1.dot(h);

        if a.abs() < 1e-12 {
            return None;
        }

        let f = 1.0 / a;
        let s = origin - self.v0;
        let u = f * s.dot(h);
        if u < 0.0 || u > 1.0 {
            return None;
        }

        let q = s.cross(edge1);
        let v = f * dir.dot(q);
        if v < 0.0 || u + v > 1.0 {
            return None;
        }

        let t = f * edge2.dot(q);
        if t < t_min || t > t_max {
            return None;
        }

        let p = origin + dir * t;
        let normal = edge1.cross(edge2).normalize();
        let mut hit = Hit { t, point: p, ..Default::default() };
        hit.set_face_normal(dir, normal);
        Some(hit)
    }
}

impl Shape for Triangle {
    fn intersect(&self, origin: Vec3, dir: Vec3, t_min: f64, t_max: f64) -> Option<Hit> {
        self.intersect_tri(origin, dir, t_min, t_max)
    }
}
